test: refactor to allow benchmark Vote (WIP, MVP)

altergui opened the pull request:
started working on #965 and opted for a refactor that will allow me reuse most of the code

altergui committed:
test: refactor to allow benchmark Vote (WIP, half-baked)

altergui committed:
mvp

altergui commented on vochain/indexer/vote.go:
```diff
@@ -41,6 +41,7 @@ func (idx *Indexer) GetEnvelope(nullifier []byte) (*indexertypes.EnvelopePackage
 		OverwriteCount:       uint32(voteRef.OverwriteCount),
 		Date:                 voteRef.BlockTime,
 		Meta: indexertypes.EnvelopeMetadata{
+			VoterID:   voteRef.VoterID.Address(),
```
this is actually a hotfix of an issue i found out while working on this. will push a separate PR for this

altergui commented on test/api_test.go:
```diff
@@ -138,51 +186,54 @@ func TestAPIcensusAndVote(t *testing.T) {
+	}
 }
 
 func TestAPIaccount(t *testing.T) {
```
there's a lot of code in TestAPIaccount that could be deduplicated now, after the proposed refactoring. i'd include this in the same PR

p4u commented on test/benchmark_vochain_test.go:
```diff
@@ -0,0 +1,26 @@
+package test
```
@mvdan can you check the basics of this benchmark and maybe propose improvements? The idea is to have a full benchmark to test performance of state+app+indexer+api

mvdan replied:
I think the logic looks okay. Two suggestions come to mind:

* What kind of election is this - encrypted or not? We should probably benchmark the one that is most common. If both are common, then I would suggest that the benchmark do _both_ side by side. We could similarly think about other kinds of votes, like overwrites.

* You use a sequential benchmark via `b.N`, which is fine, but not very realistic - in reality, people won't vote in a big election in a big line :) For that reason, I would use a parallel benchmark via https://pkg.go.dev/testing#B.RunParallel. You can still use `b.N` as the total number of iterations to set up the census and anything else you need, but the looping should use https://pkg.go.dev/testing#PB.Next instead.

mvdan replied:
Also, should the benchmark advance votes while iterating? Because again, in reality, you won't have an entire election get all of its votes within a single block. Plus, the indexer does a significant amount of its work in the `Commit` method, so you should advance at least one block while the benchmark timer is on.

altergui committed:
ci: run benchmarks (job_go_bench)

altergui committed:
use b.Run()

altergui committed:
api_test wip

altergui committed:
mvp using b.Timer in different Benchmarks

altergui committed:
arbo: remove failing BenchmarkAdd

altergui committed:
vochain: remove failing BenchmarkCheckTx

it's failing with:
ERR vochain/app.go:287 > checkTx error="voteTx: maxCensusSize reached 0/0"
app_benchmark_test.go:131: checkTX failed: checkTx voteTx: maxCensusSize reached 0/0

altergui committed:
indexer: remove failing BenchmarkFetchTx

it's failing with:
--- FAIL: BenchmarkFetchTx
    bench_test.go:157:
        error:
          got non-nil error
        got:
          e"transaction not found"
        stack:
          /home/runner/work/vocdoni-node/vocdoni-node/vochain/indexer/bench_test.go:157
            qt.Assert(b, err, qt.IsNil)
